Automattic/portkey-android
# Introducing Frame Saving process into a Service

Closes #273 

### Intro
This PR takes the former work done in order to implement a Story Frames Save architecture (see #257 and related PRs) and encapsulate the work into a Service.

This is needed given sometimes, saving a Story's frames may take time (especially if they have video-background in one or more of their Story frames), and we don't want to block the user on a screen, waiting for things to get processed.

### Services
The Android documentation describes a few of the possibilities we have for performing [background work](https://developer.android.com/guide/background). We're going to use a Foreground Service given that's the mechanism that best fits the description of our use case as explained here.

> For user-initiated work that need to run immediately and must execute to completion, use a foreground service. Using a foreground service tells the system that the app is doing something important and it shouldn’t be killed. Foreground services are visible to users via a non-dismissible notification in the notification tray.

In our case, the user taps on `Save` button and makes it explicit that they want their Story and Story frames to get saved. Hence, it makes sense to encapsulate this work in a Foreground Service.

In our specific case, we can roughly say there are 2 big parts to finishing a Story:
1. Saving the composed Story frames to disk so these are available for upload
2. upload the frames and Story to a server.

For now, this PR will start with part 1, and move the already existing code from being carelessly called from the Activity (in the click listener for the `Next` button) into a foreground Service, as this operation can potentially be lengthy depending on hardware and amount of frames to be saved. 
For the upload part we're going to delegate that to the existing WordPress Android `UploadService`, which will manage both the adding of images and uploading a custom typed Post to an existing site (to be covered in further PRs when integration of the functionality into WPAndroid is planned).

### How it works
There's a problem I found while doing this: a [started Service](https://developer.android.com/guide/components/services#CreatingAService) is passed parameters with an Intent extras bundle, and our parameters here would be a list of images with their added Views, so unless we access the `ViewModel`'s `Repository` from the `Service` (with the Repository actually persisting the data, which is not our case right now as we just keep these in memory) then we can't pass the frames with their added views if we don't first serialize this information (that is, positioning, rotation and scale information for each of the added views). 

Given our Service is actually triggered from an Activity and is quite related to it (it's a user initiated action), looks like this is a good candidate for a [bound service](https://developer.android.com/guide/components/bound-services#Creating), thus allowing us to bypass the serialization process which among other things would also incur into some time penalty (both for execution and development time). In this case, for the reasons outlined above we also need to making it both a bound Service _and_ a foreground service (as we need the background saving process to exist even if the user navigates away from the Activity or even further, away from the app itself). This PR only tackled the bound Service part, and leaves the foreground Service for an upcoming PR so to keep things simpler and easier to review.

Once the processing is ready, it can either have a successful ending, or some of the frames may have not been saved successfully. The error collection strategy will be tackled on another PR, for now we're introducing EventBus here just as a means to signal the saving process has ended (either successfully or not, it doesn't matter for now in this PR), so the event listener can properly do the cleanup they want.

### Next steps (to be tackled in subsequent PRs):
- [x] implement foreground service (with foreground notification) https://github.com/Automattic/portkey-android/pull/279
- [ ] implement error collection strategy
- [ ] implement error collection UI

### A word on future WPAndroid integration
Before tackling this PR, I studied how other services were implemented, particularly [SiteCreationService](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/ui/sitecreation/services/SiteCreationService.kt) as a model for a well written Service using the latest guidelines and good stuff Kotlin brings (such as corroutines) and the [UploadService](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/ui/uploads/UploadService.java) given our process output here will be fed into the UploadService afterwards. We followed the same pattern as used for other Services in WPAndroid, which will make it easier later when we integrate this functionality there:
- the new class extends Service
- we make it a foreground Service, like the UploadService is. Although a Story can be saved in just a few seconds (while this depends, it's generally less than 2 seconds for ~14 frames as per our tests), we could also improve this part later when integrating with WPAndroid and make use of `AutoForeground` which is implemented there for the `SiteCreationService`.


#### To test:
1. open portkey
2. tap on the + FAB
3. either capture pictures or hold the big button on the bottom to start recording a video, several times
4. add some text / emoji to your frames (this makes processing harder, just 1 added view should be fine for this PR purposes)
5. once you're ready, tap NEXT
6. observe a "Saving..." scrim appears (navigating away from this will probably break stuff as this is just introducing a bound Service, so please just stay there for now :) )
7. once the process ends, the "Saving" scrim disappears
8. got to the Photos app and find your newly created frames (videos, pictures) with your added text / emoji there.
